src/docs/widgets.md

```diff
@@ -0,0 +1,157 @@
+</ul>
+
+### Implementing a widget
+
+For the implementation of custom widgets, Theia provides several base classes to inherit from. This enables you to focus on creating your custom UI only, as the base classes already implement most required functions. Theia does not depend on a specific UI technology, like React, Vue.js or Angular. However, it provides convenience support by providing respective base classes, e.g. for React. If in doubt, using React is the most common choice for implementing custom widgets at the moment. Below you find an excerpt of the class hierarchy. In case you want to implement a widget using react, choose `ReactWidget` as a base class. If you want to implement a widget that mainly displays a tree, use `TreeWidget`. If you do not want to use React, use `BaseWidget`. Browse the type hierarchy of `BaseWidget` to see other available options.
+
+<ul>
+<li>BaseWidget
+    <ul>
+    <li><b>ReactWidget</b>
+        <ul>
+        <li>TreeWidget</li>
+        <li>…</li>
+        </ul>
+    </li>
+    </ul>
+</li>
+</ul>
+
+In the following code examples, we use `ReactWidget` as a base class. As shown below, we first initialize the widget with some base parameters:
+
+* `id`: To uniquely identify the widget, e.g. to open it via the WidgetManager.
+* `label`: Shown in the tab when the widget is open.
+* `caption`: Shown when hovering over the tab when the widget is open.
+* `closable`: Whether the user can close the widget (via the “x” in the tab or via right click menu).
+* `iconClass`: The icon shown in the tab when the widget is opened.
+
+**mywidget-widget.ts**
+```typescript
+@injectable()
+export class MyWidget extends ReactWidget {
+
+static readonly ID = 'my:widget';
+static readonly LABEL = 'My Widget';
+
+@postConstruct()
+protected async init(): Promise < void> {
+    this.id = MyWidget.ID;
+    this.title.label = MyWidget.LABEL;
+    this.title.caption = MyWidget.LABEL;
+    this.title.closable = true;
+    this.title.iconClass = 'fa fa-window-maximize'; // example widget icon.
+    this.update();
+}
+```
+
+When using the respective base class, the implementation of a widget can be really minimal and focussed on the custom UI part. In our example, we just implement the render function that will create our actual custom UI (using JSX/React). The example UI contains a button that will trigger the `displayMessage` function below.
+
+**mywidget-widget.ts**
+```typescript
+protected render(): React.ReactNode {
+    const header = `This is a sample widget which simply calls the messageService in order to display an info message to end users.`;
+    return <div id='widget-container'>
+              <AlertMessage type='INFO' header={header} />
+              <button className='theia-button secondary' title='Display Message' onClick={_a => this.displayMessage()}>Display Message</button>
+           </div>
+}
+
+@inject(MessageService)
+protected readonly messageService!: MessageService;
```

Comment:
Any reason we add the injection later? It does not follow the generator, or general practice.
It might be better to simply add it to the code snippet above instead, or not break up the code for the widget.

Reply:
Separation of concerns. If we add it above, we need to describe it above, but I think it is better described here in the context of its usage.

Reply:
OK, this is clearly a matter of taste. I personally prefer the distinct steps (following Thomas argument). At the same time, I would like to stay close to the template when it does not hurt (which I think it does not in this case). Therefore, I consider the current state a good compromise between both directions.